### PR TITLE
convert instances of memcmp to kj::ArrayPtr equality, where appropriate

### PR DESCRIPTION
**Pull request by AdityaAtulTewari**

### CODE REVIEW

File: c++/src/kj/async-io-test.c++ [hunk 1]
```diff
@@ -1294,7 +1294,7 @@ kj::Promise<void> expectRead(kj::AsyncInputStream& in, kj::StringPtr expected) {
     }
 
     auto actual = buffer.first(amount);
-    if (memcmp(actual.begin(), expected.begin(), actual.size()) != 0) {
+    if (actual != expected.first(amount)) {
```
Comment (mikea) [inline c++/src/kj/async-io-test.c++:1297]: btw there's also ArrayPtr::startsWith which might be convenient too.

File: c++/src/kj/debug.c++ [hunk 2]
```diff
@@ -287,7 +287,7 @@ static String makeDescriptionImpl(DescriptionStyle style, const char* code, int
 
     auto needsLabel = [](ArrayPtr<const char> &argName) -> bool {
       return (argName.size() > 0 && argName[0] != '\"' &&
-          !(argName.size() >= 8 && memcmp(argName.begin(), "kj::str(", 8) == 0));
+          !(argName.size() >= 8 && argName.first(8) == "kj::str("_kjb));
```
Comment (mikea) [inline c++/src/kj/debug.c++:290]: try startsWith?

File: c++/src/capnp/any.c++ [hunk 3]
```diff
@@ -92,12 +92,8 @@ Equality AnyStruct::Reader::equals(AnyStruct::Reader right) const {
     -- dataSizeR;
   }
 
-  if(dataSizeL != dataSizeR) {
-    return Equality::NOT_EQUAL;
-  }
-
-  if(0 != memcmp(dataL.begin(), dataR.begin(), dataSizeL)) {
-    return Equality::NOT_EQUAL;
+  if(dataL.first(dataSizeL) != dataR.first(dataSizeR)) {
```
[the remark this answers is not in the document]
Comment (AdityaAtulTewari) [reply c++/src/capnp/any.c++:95]: That's what I thought at first, but the buffers are allocated much larger then subtracted.

File: c++/src/kj/cidr.c++ [hunk 4]
```diff
@@ -111,7 +111,7 @@ bool CidrRange::matches(const struct sockaddr* addr) const {
       if (addr->sa_family == AF_INET6) {
         otherBits = reinterpret_cast<const struct sockaddr_in6*>(addr)->sin6_addr.s6_addr;
         static constexpr byte V6MAPPED[12] = { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0xff, 0xff };
-        if (memcmp(otherBits, V6MAPPED, sizeof(V6MAPPED)) == 0) {
+        if (kj::arrayPtr(otherBits, sizeof(V6MAPPED)) == V6MAPPED) {
```
[the remark this answers is not in the document]
Comment (AdityaAtulTewari) [reply c++/src/kj/cidr.c++:114]: They aren't the same at least on unix it's 16 vs 12 bytes.

File: c++/src/capnp/any.c++ (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikea) [reply c++/src/capnp/any.c++:95]: but the original code checks for length equality but yours doesn't! the behavior is different now.

File: c++/src/kj/cidr.c++ (hunk 4 above)
[the remark this answers is not in the document]
Comment (mikea) [reply c++/src/kj/cidr.c++:114]: let's do <= then? and also `kj::arrayPtr(otherBits).startsWith(V6MAPPED)`

File: c++/src/capnp/compiler/capnpc-c++.c++ [hunk 5]
```diff
@@ -33,6 +33,7 @@
 #include <kj/tuple.h>
 #include <kj/vector.h>
 #include <kj/filesystem.h>
+#include <kj/common.h>
```
Comment (mikea) [inline c++/src/capnp/compiler/capnpc-c++.c++:36]: I'm pretty sure common is already included everywhere. no need.

File: c++/src/capnp/any.c++ (hunk 3 above)
[the remark this answers is not in the document]
Comment (AdityaAtulTewari) [reply c++/src/capnp/any.c++:95]: This does check that dataSizeL and dataSizeR are the same as the first line of `operator==` does that comparison.

**Pull request closed without merging**